### PR TITLE
Make `type-parameter` optional in `future-type-descriptor` parsing

### LINKED ISSUE #28758: Cannot define a future type descriptor without the type parameter
**Description:**
The [type parameter is optional in the future type-descriptor](https://ballerina.io/spec/lang/master/#future-type-descriptor) but this doesn't seem to work now.

This change was initially introduced for the old parser in https://github.com/ballerina-platform/ballerina-lang/pull/23137/files but the tests don't seem to be running.

**Steps to reproduce:**
```ballerina
public function main() {
    future x;
}
```

```cmd
$ bal run test.bal 

Compiling source
        test.bal
ERROR [test.bal:(2:12,2:12)] missing lt token
ERROR [test.bal:(2:12,2:13)] unknown type 'x'
ERROR [test.bal:(2:13,2:13)] missing gt token
ERROR [test.bal:(2:13,2:13)] missing identifier
error: compilation contains errors
```

**Affected Versions:**
slalpha2

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/17922

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes #28758 
Fix `future-type-desc` parsing of the sample in #30325

## Approach
N/A

## Samples
N/A

## Remarks
- Renamed the existing `ParameterizedTypeDescriptorNode` to `MapTypeDescriptorNode`.
- Used `ParameterizedTypeDescriptorNode` to represent `error`,`xml`,`typedesc` and `future` type descriptors.
  (SyntaxKind is used to distinguish among 4 cases)

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MapTypeDescriptorNodeContext.java [hunk 1]
```diff
@@ -35,15 +35,15 @@
  * @since 2.0.0
  */
 @JavaSPIService("org.ballerinalang.langserver.commons.completion.spi.BallerinaCompletionProvider")
-public class ParameterizedTypeDescriptorNodeContext
-        extends AbstractCompletionProvider<ParameterizedTypeDescriptorNode> {
+public class MapTypeDescriptorNodeContext
+        extends AbstractCompletionProvider<MapTypeDescriptorNode> {
```
Comment (lochana-chathura) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MapTypeDescriptorNodeContext.java:39]: Also will create a separate LS context for the future type.

Comment (lochana-chathura) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MapTypeDescriptorNodeContext.java:39]: Resolving as this class is obsolete.

**Review by KRVPerera — DISMISSED** [review]
LGTM

**Review by shehan360 — APPROVED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/MapTypeDescriptorNode.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/MapTypeDescriptorNode.java:2]: this an auto-generated class via treegen.
Currently, the template header is hardcoded to 2020 at [external_node_template.mustache](https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-treegen/src/main/resources/external_node_template.mustache)

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 3]
```diff
@@ -968,10 +966,43 @@ public BLangNode transform(TupleTypeDescriptorNode tupleTypeDescriptorNode) {
...
+            constrainedType.type = refType;
+            constrainedType.constraint = createTypeNode(typeParam.get().typeNode());
+            constrainedType.pos = refType.pos;
+            return constrainedType;
+        }
+        
+        return refType;
+    }
+
+    private TypeKind getTypeKind(SyntaxKind syntaxKind) {
```
Comment (rdulmina) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:990]:
```suggestion
    private TypeKind getParameterizedTypeKind(SyntaxKind syntaxKind) {
```

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -8517,14 +8519,32 @@ private STNode parseRightArrow() {
...
+        return STNodeFactory.createMapTypeDescriptorNode(mapKeyword, typeParameter);
+    }
+
+    /**
+     * Parse future type descriptor.
+     * future-type-descriptor := `future` [type-parameter]
+     *
+     * @return Parsed node
+     */
+    private STNode parseFutureTypeDescriptor(STNode futureKeywordToken) {
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:8538]: Can we use a single function to parse all parameterized types?. We can get the SyntaxKind from parsed keyword

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 5]
```diff
@@ -161,10 +161,11 @@
             ParserRuleContext.SIMPLE_TYPE_DESC_IDENTIFIER,
             ParserRuleContext.TYPE_REFERENCE, ParserRuleContext.SIMPLE_TYPE_DESCRIPTOR,
             ParserRuleContext.OBJECT_TYPE_DESCRIPTOR, ParserRuleContext.RECORD_TYPE_DESCRIPTOR,
-            ParserRuleContext.PARAMETERIZED_TYPE, ParserRuleContext.TUPLE_TYPE_DESC_START,
+            ParserRuleContext.MAP_TYPE_DESCRIPTOR, ParserRuleContext.TUPLE_TYPE_DESC_START,
             ParserRuleContext.ERROR_KEYWORD, ParserRuleContext.STREAM_KEYWORD, ParserRuleContext.TABLE_KEYWORD,
             ParserRuleContext.FUNC_TYPE_DESC, ParserRuleContext.PARENTHESISED_TYPE_DESC_START,
-            ParserRuleContext.CONSTANT_EXPRESSION, ParserRuleContext.TYPEDESC_TYPE_DESCRIPTOR };
+            ParserRuleContext.CONSTANT_EXPRESSION, ParserRuleContext.TYPEDESC_TYPE_DESCRIPTOR,
+            ParserRuleContext.FUTURE_TYPE_DESCRIPTOR, ParserRuleContext.XML_TYPE_DESCRIPTOR };
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java:168]: Can we use one context for all the type descriptors?

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STNodeTransformer.java [hunk 6]
```diff
@@ -285,8 +285,8 @@ public T transform(STRemoteMethodCallActionNode remoteMethodCallActionNode) {
         return transformSyntaxNode(remoteMethodCallActionNode);
     }
 
-    public T transform(STParameterizedTypeDescriptorNode parameterizedTypeDescriptorNode) {
-        return transformSyntaxNode(parameterizedTypeDescriptorNode);
+    public T transform(STMapTypeDescriptorNode mapTypeDescriptorNode) {
+        return transformSyntaxNode(mapTypeDescriptorNode);
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STNodeTransformer.java:289]: Should we add a simple test to cover this line?

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/MapTypeDescriptorNode.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/MapTypeDescriptorNode.java:2]: Let's create an issue to update that mustache

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STNodeTransformer.java (hunk 6 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STNodeTransformer.java:289]: BLangNodetranformer is a use case of the external node transformer. Do we any use for the internal node transformer?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STNodeTransformer.java:289]: BLangNodetranformer is a use case of the external node transformer. Do we any use case of the internal node transformer?

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/MapTypeDescriptorNode.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/MapTypeDescriptorNode.java:2]: #31047

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STNodeTransformer.java (hunk 6 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STNodeTransformer.java:289]: I think currently we use it only in BallerinaParser.java. Yeh, it's not possible to cover this line without requirement.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 7]
```diff
@@ -5551,4 +5529,22 @@ private boolean isSingleKeywordAttachPointIdent(SyntaxKind tokenKind) {
...
+        switch (tokenKind) {
+            case TYPEDESC_KEYWORD:
+            case FUTURE_KEYWORD:
+            case XML_KEYWORD:
+            case ERROR_KEYWORD:
+                return true;
+            default:
+                return false;
+        }
+    }
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java]: Should we move this to BallerinaParser.java and use this method there as well as we did for `isPredeclaredPrefix` method?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_positive.bal [hunk 8]
```diff
@@ -121,13 +123,14 @@ function testCustomErrorFuture() {
...
-//
-//    any|error x = wait te;
-//
-//    assertEquality(getError().toString(), x.toString());
-//}
+function testCustomErrorFutureWithoutConstraint() {
+    future te = start getError();
+
+    any|error x = wait te;
+    string str = (<error>x).toString();
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_positive.bal:130]:
```suggestion
    string str = x is error ? x.toString() : x.toString()
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_positive.bal [hunk 9]
```diff
@@ -121,13 +123,14 @@ function testCustomErrorFuture() {
...
+
+    assertEquality(getError().toString(), str);
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_positive.bal:132]:
```suggestion
    assertEquality(getError(), x);
```

Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_positive.bal:132]: Will this work?

**Review by rdulmina — APPROVED** [review]
LGTM

**Pull request merged**